src/lib/snark_work_lib/partitioned.ml

```diff
@@ -0,0 +1,534 @@
+      | Merge { proof1; proof2; _ } -> (
+          let module Statement = Mina_state.Snarked_ledger_state in
+          let { Proof_carrying_data.data = t1; _ } = proof1 in
+          let { Proof_carrying_data.data = t2; _ } = proof2 in
+          let stmt =
+            Statement.merge
+              ({ t1 with sok_digest = () } : Statement.t)
+              { t2 with sok_digest = () }
+          in
+          match stmt with
+          | Ok stmt ->
+              stmt
+          | Error e ->
+              failwithf
+                "Failed to construct a statement from  zkapp merge command %s"
+                (Error.to_string_hum e) () )
+  end
+
+  [%%versioned
+  module Stable = struct
+    [@@@no_toplevel_latest_type]
+
+    module V1 = struct
+      type t =
+        { spec : Spec.Stable.V1.t
+        ; pairing : Pairing.Sub_zkapp.Stable.V1.t
+        ; job_id : ID.Stable.V1.t
+        ; common : Spec_common.Stable.V1.t
+        }
+      [@@deriving sexp, yojson]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  type t =
+    { spec : Spec.t
+    ; pairing : Pairing.Sub_zkapp.t
+    ; job_id : ID.t
+    ; common : Spec_common.t
+    }
+
+  let read_all_proofs_from_disk ({ spec; pairing; job_id; common } : t) :
+      Stable.Latest.t =
+    { spec = Spec.read_all_proofs_from_disk spec; pairing; job_id; common }
+
+  let write_all_proofs_to_disk ~(proof_cache_db : Proof_cache_tag.cache_db)
+      ({ spec; pairing; job_id; common } : Stable.Latest.t) : t =
+    { spec = Spec.write_all_proofs_to_disk ~proof_cache_db spec
+    ; pairing
+    ; job_id
+    ; common
+    }
+end
+
+module Spec = struct
+  module Poly = struct
+    [%%versioned
+    module Stable = struct
+      [@@@no_toplevel_latest_type]
+
+      module V1 = struct
+        type 'metric t =
+          | Single of
+              { single_spec : Selector.Single.Spec.Stable.V1.t
+              ; pairing : Pairing.Single.Stable.V1.t
+              ; metric : 'metric
+              ; common : Spec_common.Stable.V1.t
+              }
+          | Sub_zkapp_command of
+              { spec : Zkapp_command_job.Stable.V1.t; metric : 'metric }
+          | Old of
+              { instances :
+                  (Selector.Single.Spec.Stable.V1.t * 'metric)
+                  One_or_two.Stable.V1.t
+              ; common : Spec_common.Stable.V1.t
+              }
+        [@@deriving sexp, yojson]
+
+        let to_latest = Fn.id
+      end
+    end]
+
+    type 'metric t =
+      | Single of
+          { single_spec : Selector.Single.Spec.t
+          ; pairing : Pairing.Single.t
+          ; metric : 'metric
+          ; common : Spec_common.Stable.V1.t
+          }
+      | Sub_zkapp_command of { spec : Zkapp_command_job.t; metric : 'metric }
+      | Old of
+          { instances : (Selector.Single.Spec.t * 'metric) One_or_two.t
+          ; common : Spec_common.t
+          }
+
+    let map_metric (t : 'm t) ~(f : 'm -> 'n) : 'n t =
```

Comment:
Nit: I would simply call it `map` as the signature explains that it would impact the parameter (which contains the "metric")